### PR TITLE
Progress events without breaking pub-sub

### PR DESCRIPTION
**Pull request by Gozala**
This is second take on #52, which end up breaking pubsub. It is based on #57 which fixed another issue.

With this changes `fetch` will use XHR if either `options.onDownloadProgress` or `options.onUploadProgress` is provided. Otherwise it will stick to native `fetch`. This way `file.add` could opt-in into observing progress, but have all of response be buffered before ability to consume it. Pubsub on the other hand doesn't really care about progress events and therefor will continue using `fetch` that can have a streaming response body.

### CODE REVIEW

**Comment by Gozala** [comment]
Electron seems to be failing because we use native fetch there, while we're passing non native `AbortController`. Prior to #54 we used `node-fetch` in electron so this issue did not came up. Then #54 just switched to XHR (using native APIs in Electron). And now that we're going for hybrid approach it became a problem, because native `fetch` in electron expects native `AbortSignal` instead of custom implementation from [abort-controller](https://npmjs.com/package/abort-controller)

I think proper solution would be to make `abort-controller` use native one in electron (or wherever it's available really). But I think we might need a shortcut here.

**Comment by Gozala** [comment]
Now this got complicated. It looks like http module makes use of [any-singnal](https://github.com/jacobheun/any-signal) which in turn uses `abort-controller`. So even if native `AbortSignal` is passed it will be turned into non native. Switching any-signal to use `AbortController` from `ipfs-utils` would have been an option, but that would introduce circular dependency 😖

Not sure what's the best thing to do here is. I have created https://github.com/mysticatea/abort-controller/issues/24

**Comment by hugomrdias** [comment]
I would really like to avoid #52 #54 #57 and this one by doing nothing or at best implement download progress (which can be done using fetch). 
This is getting too complex and will only increase maintenance #54 is already and example of that. 

Upload progress is and always was a dream thats why browsers never implemented it (or streams) any box between client and origin server can buffer the request and the progress means nothing. Hopefully in the future this will change (chrome is already implementing streams but warns about this problem).

Sorry for not supporting this, lots of good work went into making this, but i really think its best not to increase complexity for the upload progress.

This is just my opinion if others support this i will not block it.

**Comment by Gozala** [comment]
> Upload progress is and always was a dream thats why browsers never implemented it (or streams) any box between client and origin server can buffer the request and the progress means nothing. Hopefully in the future this will change (chrome is already implementing streams but warns about this problem).

The main motivation here is to support upload progress in webui. Adding a large file takes a long enough that webui appears broken. E.g. 2gig file takes 10-12secs (for more details and numbers see https://github.com/ipfs-shipyard/ipfs-webui/pull/1534)

With this changes it was possible to render fairly accurate progrssbar. Your comment about box between server and client is correct, but it does not apply to the problem we were trying to solve, because webui talks to the local node.

I personally do not think that waiting on chrome to ship request streams, then waiting for it to propagate to the electron and giving up on non-chromium browsers in order to reduced complexity is the right tradeoff, because it leaves wrong impression  of the IPFS ecosystem when someone is just trying to add a file.

However there might be alternative ways we could consider to provide a feedback, e.g. infinite spinners. They do help a bit, but don't really work for long tasks that take 10secs.

**Comment by jacobheun** [comment]
I think the UX improvement this adds outweighs the additional complexity. We should add some additional tests here to catch regressions. @Gozala can we add some tests to validate this actually fixes the streaming issue we had with pubsub? 

While I did run this branch against js-ipfs previously and the tests passed, we shouldn't need to rely on that suite to verify things are working as expected.

**Comment by lidel** [comment]
> I personally do not think that waiting on chrome to ship request streams, then waiting for it to propagate to the electron and giving up on non-chromium browsers in order to reduced complexity is the right tradeoff, because it leaves wrong impression of the IPFS ecosystem when someone is just trying to add a file.

+1 on this.  I believe that shifting burden from the library to its  user is a bad trade off here:

We document `progress` handler at [`js-ipfs/docs/core-api/`](https://github.com/ipfs/js-ipfs/blob/master/docs/core-api/FILES.md#options) and it is a terrible devexp when it really does not work over HTTP API. 

Bunch of developers try to implement progress reporting, hit this problem, and waste bunch of time debugging because our docs say `progress` should work, but it does not really work.

Even if our docs were honest and admit it does not work over HTTP API, people have the same problem we have in webui right now, and will waste time on implementing alternative UX strategies to indicate "fake progress". 

Sad reality is that nobody will care or appreciate that we "wait 2 or 5 more years for browsers to fix this mess",
devs new to the ecosystem will blame IPFS for either dishonest/outdated docs or `ipfs.add` missing "the most basic functionality".

**Comment by achingbrain** [comment]
This is the wrong solution as it ignores what's actually going during the API call (e.g. the HTTP request and the API request may have different notions of 'progress'), but sadly the right solution is completely impractical right now (bidirectional HTTP streaming in browsers) so it'll have to do.

Can you also please:

1. Open a PR against js-ipfs with this branch as the `ipfs-utils` dep version to make sure the tests pass with these changes
1. Expose and document the new upload/download progress args in ipfs-http-client
    - These probably just need documenting under [additional options](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-http-client#additional-options) in the ipfs-http-client README, not the core-api docs as they make no sense in the context of calling into core/running a full IPFS node in the browser or node.js
1. Ensure it doesn't cause a regression for https://github.com/ipfs/js-ipfs/issues/3238

**Comment by Gozala** [comment]
>   1. Open a PR against js-ipfs with this branch as the `ipfs-utils` dep version to make sure the tests pass with these changes

https://github.com/ipfs/js-ipfs/pull/3310

> 2. Expose and document the new upload/download progress args in ipfs-http-client
> 
>     * These probably just need documenting under [additional options](https://github.com/ipfs/js-ipfs/tree/master/packages/ipfs-http-client#additional-options) in the ipfs-http-client README, not the core-api docs as they make no sense in the context of calling into core/running a full IPFS node in the browser or node.js

I added jsdoc comment but have not added it the README, because I don't think we want users to pass this option as it would disable streaming. It is only used by `ipfs.add` and even there user does not need to provide additional option, existing `progress` handler is used.

> 3. Ensure it doesn't cause a regression for ipfs/js-ipfs#3238

Verified

**Comment by Gozala** [comment]
Had to reset on top of current master, otherwise github complained about non-clean merge even though I did merge master in 🤷‍♂️

**Comment by Gozala** [comment]
merged #62 into this as there was some overlap between two

**Comment by achingbrain** [comment]
Can you fix the conflicts please?  Then we should try to get this resolved.

**Comment by Gozala** [comment]
So #62 has introduced @achingbrain/electron-fetch as a dependency, which is API incompatible with both node-fetch and native fetch 😭, more specifically native and a `node-fetch` `Response` implementations accept `Uint8Array`s for a body, but electron-fetch does not:

https://github.com/achingbrain/electron-fetch/blob/b6a8cb74b4f04837140044af3396e150a5ee2e8c/src/body.js#L24-L44

Which causes them to be treated as strings leading to incorrect behavior here:

https://github.com/ipfs/js-ipfs-utils/blob/8111a80b470108b7602d034f440129a224c9125e/src/http/fetch.node.js#L82-L99

Furthermore unlike node-fetch electron-fetch represents `response.body` in different representations from node-fetch, which creates more problems
https://github.com/node-fetch/node-fetch/blob/cbd4d895767e33491219e40013dd8daa5c7ac024/src/body.js#L35-L69

What's worth both seem to ignore that `response.body` supposed be `ReadableStream`.

**Comment by Gozala** [comment]
@achingbrain I've update patch so that it overcomes issues mentioned in last comment. While it would make sense to make changes to electron-fetch to make it more API compatible with node-fetch, I would really really like to avoid blocking this further, and pursue that separately. This is already several degrees separated from the actual task of **having a progress bar when you import files in webui**.

**Review by lidel — APPROVED** [review]
As long, we have tests that guard against regressions (see comment inline), lgtm!

@achingbrain we need this to unblock improved import UX scheduled to land as part of pinning service integration:
- https://github.com/ipfs-shipyard/ipfs-webui/issues/1594
- https://github.com/ipfs-shipyard/ipfs-webui/pull/1655

File: test/http.spec.js [hunk 1]
```diff
@@ -150,4 +152,30 @@ describe('http', function () {
...
+    const request = await HTTP.post(`${process.env.ECHO_SERVER}/echo`, {
+      body,
+      onUploadProgress: (progress) => {
+        expect(progress).to.have.property('lengthComputable').to.be.a('boolean')
+        expect(progress).to.have.property('total', body.byteLength)
+        expect(progress).to.have.property('loaded').to.be.greaterThan(0)
+        upload += 1
+      }
+    })
+    await all(request.iterator())
```
[the remark this answers is not in the document]
Comment (Gozala) [reply test/http.spec.js:168]: Sure, I'll add that

File: test/http.spec.js [hunk 2]
```diff
@@ -150,4 +152,30 @@ describe('http', function () {
...
+
+  it('progress events', async () => {
+    let upload = 0
+    const body = new Uint8Array(1000000 / 2)
+    const request = await HTTP.post(`${process.env.ECHO_SERVER}/echo`, {
+      body,
+      onUploadProgress: (progress) => {
+        expect(progress).to.have.property('lengthComputable').to.be.a('boolean')
+        expect(progress).to.have.property('total', body.byteLength)
+        expect(progress).to.have.property('loaded').to.be.greaterThan(0)
```
Comment (achingbrain) [inline test/http.spec.js:164]: Minor nit, a bit more readable:

```suggestion
        expect(progress).to.have.property('loaded').that.is.greaterThan(0)
```

**Review by achingbrain — CHANGES_REQUESTED** [review]

File: src/http/fetch.browser.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  if (options.onUploadProgress) {
+    request.upload.onprogress = options.onUploadProgress
+  }
+
+  // Note: Need to use `arraybuffer` here instead of `blob` because `Blob`
+  // instances coming from JSDOM are not compatible with `Response` from
+  // node-fetch (which is the setup we get when testing with jest because
+  // it uses JSDOM which does not provide a global fetch
```
[the remark this answers is not in the document]
Comment (Gozala) [reply src/http/fetch.browser.js:52]: You told me in the earlier comment to:

> Ensure it doesn't cause a regression for ipfs/js-ipfs#3238

Which is why this done here.

File: test/http.spec.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gozala) [reply test/http.spec.js]: Addressed

**Comment by Gozala** [comment]
@achingbrain can you please let me know what else needs to be done here ?

**Review by achingbrain — APPROVED** [review]

**Comment by achingbrain** [comment]
@hugomrdias could you eyeball these changes and release if you're happy with them please?

**Comment by hugomrdias** [comment]
Sure I will do it asap.

Hugo Dias

**Pull request merged**